Support Windows and npm link install scenario

ayan4m1 opened the pull request:
I would like to use this package to help manage the pain point of installing `react-dom` manually. This package would seem to be the perfect solution. However, I found (after several hours investigation) that it does not seem to support Windows, because the matcher regex used to find npm and/or yarn only matches on Unix-style paths.

In testing/reproducing this, I also found that it does not work when used as an `npm link` package - because `__dirname` used in `install.js` points to the physical path on disk of the non-symlinked package directory, not the path that `install-peers` occupies under the `node_modules` of the package (e.g. `/code/install-peers` rather than `/code/install-peers-test/node_modules/install-peers` which is a symlink to `/code/install-peers`).

I am using [these configurations](https://gist.github.com/ayan4m1/e8261cb6a33a1436f8b9d0f4e68e8753) to test peer dependency installation.

Example output using 1.0.2 on Windows (`react-dom` is not present in `node_modules` afterward):

> $ npm install
> 
> > install-peers@1.0.2 postinstall C:\code\install-peers-test\node_modules\install-peers
> > node install.js
> 
> npm WARN install-peers-test@0.1.0 requires a peer of react-dom@^16.6.0 but none is installed. You must install peer dependencies yourself.
> npm WARN install-peers-test@0.1.0 No description
> npm WARN install-peers-test@0.1.0 No repository field.
> 
> added 10 packages from 4 contributors and audited 16 packages in 4.096s
> found 0 vulnerabilities

Example output using my branch on Windows (`react-dom` is present in `node_modules` afterward):

> $ npm install
> 
> > install-peers@1.0.2 postinstall C:\code\install-peers-test\node_modules\install-peers
> > node install.js
> 
> npm WARN install-peers-test@0.1.0 No description
> npm WARN install-peers-test@0.1.0 No repository field.
> 
> + react-dom@16.6.0
> added 1 package and audited 16 packages in 51.138s
> found 0 vulnerabilities
> 
> Installed peerDependencies as devDependencies via npm.
> npm notice created a lockfile as package-lock.json. You should commit this file.
> npm WARN install-peers-test@0.1.0 requires a peer of react-dom@^16.6.0 but none is installed. You must install peer dependencies yourself.
> npm WARN install-peers-test@0.1.0 No description
> npm WARN install-peers-test@0.1.0 No repository field.
> 
> added 10 packages from 4 contributors and audited 16 packages in 70.714s
> found 0 vulnerabilities

I have prepared this pull request with four changes:

* Added code to `install-yarn.js` and `install-npm.js` that "flips" the path separators if the platform is win32, allowing the regex matcher to work on Windows
* `rootPath` will be set to `process.env.INIT_CWD` if it is present (npm 5+), which allows the package to be used in `npm link` scenarios on *nix - this environment variable avoids the need to use relative pathing to find the package root
* Changed `__dirname` to `process.cwd()` in `rootPath` fallback initialization, which allows the package to be used in `npm link` scenarios
* Added a `console.error` message if neither `yarn` nor `npm` are found on the system, thus alerting the user to the reason no action was taken during execution

There is one case, where the user is on *nix and does not have npm 5+, where the `npm link` scenario will fail. However, this has always been the case. Unfortunately, `process.cwd()` `process.env.PWD` and `__dirname` are all the "incorrect" value in this environment, and `process.env.INIT_CWD` is not available.

My testing on Windows was done using Windows 10, Node 10.12.0. I regression-tested the package using it in both `npm link` and non-link scenarios on CentOS 7, Node 6.14.4 (npm 3) and 10.13.0 (npm 6).

Please let me know your thoughts and feedback.

alexindigo commented:
Hey @ayan4m1, thanks for your interest in the package.
And windows support will definitely help.

Let me go over your changes to better understand it.

In the same time, I've found that it's easier and seems to be more stable to just use cli "interface" for npm (similar way how it's done for yarn) rather than trying to keep up with internal changes.

This is how I'd like to do it in this package as well:
https://github.com/alexindigo/install-peers-cli/blob/master/install-npm.js#L9-L29

Please review that part to see if it matches your use case and how it would affect your changes in this PR.

Thank you.

alexindigo commented on install-npm.js:
```diff
@@ -4,8 +4,14 @@ var npmModule
-  npmModule = require(path.resolve(process.env['npm_execpath'], '..', '..'));
+if (process.env['npm_execpath']) {
+  if (process.platform === 'win32') {
+    process.env['npm_execpath'] = process.env['npm_execpath'].replace(/\\/g, '/');
```
I think we can do better than to mangle environment variables with something that will break downstream functionality.

ayan4m1 replied:
Done, no more changes to env vars.

alexindigo commented on install-npm.js:
```diff
@@ -4,8 +4,14 @@ var npmModule
+    process.env['npm_execpath'] = process.env['npm_execpath'].replace(/\\/g, '/');
+  }
+
+  if (process.env['npm_execpath'].match(/\/node_modules\/npm\/bin\/npm-cli\.js$/)) {
```
I think we can utilize built-in node functionality in handling platform FS differences
```suggestion
  var expectedNpmPath = path.join('bin', 'npm-cli.js')
  if (process.env['npm_execpath'].substr(-1 * expectedNpmPath.length) === expectedNpmPath) {
```

ayan4m1 replied:
Latest approach uses path.join to build a path to both NPM and Yarn binaries, then compares to execpath.

alexindigo commented on install-yarn.js:
```diff
@@ -6,8 +6,14 @@ var yarnModule
+    process.env['npm_execpath'] = process.env['npm_execpath'].replace(/\\/g, '/');
+  }
+
+  if (process.env['npm_execpath'].match(/node_modules\/yarn\/bin\/yarn\.js$/)) {
```
Same goes here, we can rely on node to do the groundwork:
```suggestion
    var expectedYarnPath = path.join('bin', 'yarn.js')
    if (process.env['npm_execpath'].substr(-1 * expectedYarnPath.length) === expectedYarnPath) {
```

ayan4m1 replied:
Latest approach uses path.join to build a path to both NPM and Yarn binaries, then compares to execpath.

alexindigo commented on install.js:
```diff
@@ -56,6 +56,8 @@ function installPeerDeps() {
     } else if (installNpm) {
       installNpm(peerDeps, peerInstallOptions, installDone.bind(null, 'npm'));
+    } else {
+      console.error('Did not find a viable package manager to install dependencies with.');
```
👌

ayan4m1 commented:
Thank you for the quick feedback! I have hopefully addressed everything you pointed out.

* Instead of mangling the env var, use `path.join` as suggested combined with slightly simpler string comparison
* Changed `install-npm` to use the executioner style invocation (taken from install-peers-cli)
* Added console.log of output for both npm and yarn invocations via executioner (taken from [install-peers-cli](https://github.com/alexindigo/install-peers-cli/blob/master/install.js#L73))
* Wrap `${node}` and `${npm/yarn}` in executioner string in double-quotes because the path can/will have un-escaped space(s) on Windows

It is worth noting that with Yarn, the console.log output is being hidden or lost due to https://github.com/yarnpkg/yarn/issues/5476 - no idea what's going on there but it threw me for a loop at first. Theoretically this code will work if and when the yarn behavior changes.

Re-tested on Windows Node 10, Linux Node 6 and 10, yarn and npm - all scenarios result in peer dependencies correctly installed.

alexindigo commented on install-npm.js:
```diff
@@ -1,27 +1,42 @@
 
-  if (process.env['npm_execpath'].match(/\/node_modules\/npm\/bin\/npm-cli\.js$/)) {
-    npmModule = require(path.resolve(process.env['npm_execpath'], '..', '..'));
+  if (execPath.slice(-expectedPath.length) === expectedPath) {
```
Do you mind to make it more explicit `-1 * expectedPath.length` ?

alexindigo commented on install-yarn.js:
```diff
@@ -1,45 +1,43 @@
 
-  if (process.env['npm_execpath'].match(/node_modules\/yarn\/bin\/yarn\.js$/)) {
-    yarnModule = path.resolve(process.env['npm_execpath'], '..', '..', 'lib', 'cli');
+  if (execPath.slice(-expectedPath.length) === expectedPath) {
```
Same here.

alexindigo commented:
Thanks a lot for the changes. I'll add your as a contributor, so you can merge when it's all clear.
And what is your npm handle, I'll add you to the project, so you can publish updated versions.

Thank you.

ayan4m1 commented:
Happy to help, thanks for being so responsive. I am also `ayan4m1` on npmjs.com. I will get some friends to test on Macs and in some virtualized environments, but ultimately I'd like to release this as 1.0.3 if that makes sense to you.

alexindigo commented:
Perfect. Thank you.

Added you on npm:

```
$ npm owner add ayan4m1 install-peers
+ ayan4m1 (install-peers)
```
